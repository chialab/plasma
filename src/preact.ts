import { mkdir, writeFile } from 'node:fs/promises';
import { dirname, join } from 'node:path';
import { filterPublicMemebers, isOptionalClassField } from './utils';
import type { Entry } from './walker';

export interface PreactTransformOptions {
    /**
     * The output directory to write the converted components to.
     */
    outdir: string;
}

function capitalize(str: string) {
    return str.charAt(0).toUpperCase() + str.slice(1);
}

export function generatePreactComponent(entry: Entry) {
    const { packageJson, definition, declaration } = entry;

    const props = filterPublicMemebers(declaration).map((member) => member.name);
    const eventProps =
        declaration.events?.reduce(
            (acc, event) => ({
                ...acc,
                [`on${capitalize(event.name)}`]: event.name,
            }),
            {} as Record<string, string>
        ) ?? {};

    return `import { h } from 'preact';
import { useRef, useEffect } from 'preact/hooks';
import '${packageJson.name}';

const properties = ${JSON.stringify(props)};
const events = ${JSON.stringify(eventProps)};

export const ${declaration.name} = ({ children, ...props }) => {
    const ref = useRef();
    const {
        ${props.join(',\n        ')},
        ${Object.keys(eventProps).join(',\n        ')},
        ...restProps
    } = props;

    useEffect(() => {
        const listeners = [];

        for (const prop in props) {
            if (prop in events) {
                ref.current.addEventListener(events[prop], props[prop]);
                listeners.push(() => ref.current.removeEventListener(events[prop], props[prop]));
            } else if (properties.includes(prop)) {
                ref.current[prop] = props[prop];
            }
        }

        return () => {
            listeners.forEach((listener) => listener());
        };
    });

    return h('${definition.extend ?? definition.name}', {
        ref,${definition.extend ? `\n        is: '${definition.name}',` : ''}
        ...restProps,
    }, children);
};`;
}

export function generatePreactTypings(entry: Entry) {
    const { packageJson, declaration } = entry;

    const imports = `import { FunctionComponent, JSX } from 'preact';
import { ${declaration.name} as Base${declaration.name} } from '${packageJson.name}';
`;

    const propertiesTypings = filterPublicMemebers(declaration).map(
        (member) =>
            `${member.name}${isOptionalClassField(member) ? '?' : ''}: Base${declaration.name}['${member.name}'];`
    );

    const eventTypings =
        declaration.events?.map(
            (event) => `on${capitalize(event.name)}?: (this: Base${declaration.name}, event: CustomEvent) => boolean;`
        ) ?? [];

    return `${imports}
export const ${declaration.name}: FunctionComponent<JSX.HTMLAttributes<Base${declaration.name}> & {
    ${propertiesTypings.join('\n    ')}
    ${eventTypings.join('\n    ')}
    [key: \`data-\${string}\`]: any;
}>;`;
}

export async function transformPreact(entry: Entry, options: PreactTransformOptions) {
    const { declaration } = entry;
    const outFile = join(options.outdir, `${declaration.name}.js`);
    const declFile = join(options.outdir, `${declaration.name}.d.ts`);

    await mkdir(dirname(outFile), {
        recursive: true,
    });
    await Promise.all([
        writeFile(outFile, generatePreactComponent(entry)),
        writeFile(declFile, generatePreactTypings(entry)),
    ]);

    return outFile;
}
