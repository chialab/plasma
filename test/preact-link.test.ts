import { h, render } from 'preact';
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { TestLink } from './src/preact/TestLink';

let host: HTMLElement;

const awaitReactRender = () => new Promise((resolve) => setTimeout(resolve, 100));

describe('Element', () => {
    beforeEach(() => {
        host = document.createElement('div');
        document.body.appendChild(host);
    });

    afterEach(() => {
        host.remove();
    });

    test('mount component', async () => {
        const onStringChange = vi.fn();
        const onClick = vi.fn((event) => event.preventDefault());
        render(
            h(TestLink, {
                'stringProp': 'test',
                'booleanProp': true,
                'numericProp': 1,
                'objectProp': { test: true },
                'data-attr': 'test',
            }),
            host
        );
        await awaitReactRender();
        const node = host.querySelector('a') as HTMLElement;
        expect(host.innerHTML).toMatchSnapshot();
        expect(onStringChange).not.toHaveBeenCalled();
        expect(onClick).not.toHaveBeenCalled();
        render(
            h(TestLink, {
                'onStringchange': onStringChange,
                onClick,
                'stringProp': 'changed',
                'booleanProp': true,
                'numericProp': 1,
                'objectProp': { test: true },
                'data-attr': 'test',
            }),
            host
        );
        await awaitReactRender();
        expect(host.innerHTML).toMatchSnapshot();
        expect(onStringChange).toHaveBeenCalled();
        node.click();
        expect(onClick).toHaveBeenCalled();
    });
});
