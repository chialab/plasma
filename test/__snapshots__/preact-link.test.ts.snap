// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Element > mount component 1`] = `"<a is=\\"test-link\\" data-attr=\\"test\\" :scope=\\"test-link\\" :defined=\\"\\" stringprop=\\"test\\" booleanprop=\\"\\" numericprop=\\"1\\"><span class=\\"text\\">test</span><span class=\\"boolean\\">true</span><span class=\\"number\\">1</span><span class=\\"object\\">{\\"test\\":true}</span></a>"`;

exports[`Element > mount component 2`] = `"<a is=\\"test-link\\" data-attr=\\"test\\" :scope=\\"test-link\\" :defined=\\"\\" stringprop=\\"changed\\" booleanprop=\\"\\" numericprop=\\"1\\"><span class=\\"text\\">changed</span><span class=\\"boolean\\">true</span><span class=\\"number\\">1</span><span class=\\"object\\">{\\"test\\":true}</span></a>"`;
